### PR TITLE
dotnet 7.0.3 and 6.0.14

### PR DESCRIPTION
**Pull request by kasperk81**
https://devblogs.microsoft.com/dotnet/february-2023-updates/

### CODE REVIEW

**Comment by kasperk81** [comment]
@mattgodbolt assuming these are not set to automatic builds (unlike trunk), please schedule builds for v6 and v7 after this is merged. thanks!

**Comment by mattgodbolt** [comment]
;We typically don't delete old versions: if we do then old links that refer to the previous compilers will break.

We either:
* accept this
* add aliases in the configuration file (wihch means we silently point people at the new version, which might be surprising)
* add the new compiler and "hide" the old one which means it can no longer be selected but old links will work
* just add new compiles and leave the old one

For C++ we prefer the last as it's the least surprising and means folks can bisect compiler issues. But I'm not sure what is best for dotnet.

**Comment by mattgodbolt** [comment]
(apologies if we've had this conversation before and I forgot the outcome...)

**Comment by mattgodbolt** [comment]
Builds:
* https://github.com/compiler-explorer/infra/actions/runs/4189921860
* https://github.com/compiler-explorer/infra/actions/runs/4189923185

**Comment by mattgodbolt** [comment]
Both the builds failed as apparently I need to put `v` in the prefix. I don't know how this worked before

**Comment by mattgodbolt** [comment]
new builds: 
* https://github.com/compiler-explorer/infra/actions/runs/4190000817
* https://github.com/compiler-explorer/infra/actions/runs/4189999700

**Comment by kasperk81** [comment]
yea, their labels are v-prefixed.

> add the new compiler and "hide" the old one which means it can no longer be selected but old links will work

i think this is better. can you point me to an example where we hide it? i am guessing that doesn't work for `trunk` links.

**Comment by mattgodbolt** [comment]
> yea, their labels are v-prefixed.
> 
> > add the new compiler and "hide" the old one which means it can no longer be selected but old links will work
> 
> i think this is better. can you point me to an example where we hide it? i am guessing that doesn't work for `trunk` links.

Trunk is different: we _do_ replace trunk in-place, so there's never a need to alias or hide them.

Some examples:
```
$ git grep hidden etc
etc/config/c++.amazon.properties:compiler.gcc-embed-trunk.hidden=true
etc/config/c++.amazon.properties:compiler.gcc-static-analysis-trunk.hidden=true
etc/config/c++.amazon.properties:compiler.clang_embed.hidden=true
etc/config/c++.amazon.properties:compiler.k1cg741.hidden=true
etc/config/c++.amazon.properties:compiler.k1cg750.hidden=true
etc/config/c.amazon.properties:compiler.cgstatic-analysis.hidden=true
etc/config/cuda.amazon.properties:compiler.hiptrunk.hidden=true
etc/config/d.amazon.properties:compiler.gdc49.hidden=true
etc/config/d.amazon.properties:compiler.gdc48.hidden=true
etc/config/d.amazon.properties:compiler.gdc52.hidden=true
```

(hidden will not be shown but is available. The important part is the ID must remain the same, and the updated compilers need new IDs)

File: bin/yaml/dotnet.yaml [hunk 1]
```diff
@@ -8,7 +8,7 @@ compilers:
       - name: trunk
         if: nightly
         type: nightly
-      - v7.0.1
-      - name: v6.0.11
+      - v7.0.3
+      - name: v6.0.14
         check_env:
           - DOTNET_CLI_HOME=/tmp
```
Comment (kasperk81) [inline bin/yaml/dotnet.yaml:14]:
```suggestion
      - v6.0.11
      - v6.0.14
      - v7.0.1
      - v7.0.3
```

**Comment by kasperk81** [comment]
changed to add new versions. i'll update configs https://github.com/compiler-explorer/compiler-explorer/blob/main/etc/config/csharp.amazon.properties and others once new images are built

File: bin/yaml/dotnet.yaml [hunk 2]
```diff
@@ -8,7 +8,7 @@ compilers:
       - name: trunk
         if: nightly
         type: nightly
+      - v6.0.11
+      - v6.0.14
       - v7.0.1
-      - name: v6.0.11
```
Comment (mattgodbolt) [inline bin/yaml/dotnet.yaml:12]: Does 6.0.11 no longer need this environment variable set to run?

Comment (kasperk81) [reply bin/yaml/dotnet.yaml:12]: it is now set in https://github.com/compiler-explorer/compiler-explorer/blob/60d2f8fa5f02392be84990bcf947aae090d01303/lib/compilers/dotnet.ts

Comment (partouf) [reply bin/yaml/dotnet.yaml:12]: This yaml does NOT use that compiler-explorer script, the `ce_install` command requires the `--version` to see if the dotnet executable is there, are you sure that will work without the environment variable? have you tested it?

Comment (kasperk81) [reply bin/yaml/dotnet.yaml:12]: > have you tested it?

no but ci did https://github.com/compiler-explorer/infra/actions/runs/4195904213 and it's green.

Comment (partouf) [reply bin/yaml/dotnet.yaml:12]: The CI cannot test that. You can test it by following the instructions on the readme https://github.com/compiler-explorer/infra/blob/main/README.md

Comment (kasperk81) [reply bin/yaml/dotnet.yaml:12]: tested, dotnet --version is working without this environment variable

Comment (partouf) [reply bin/yaml/dotnet.yaml:12]: Cool, thanks

**Comment by mattgodbolt** [comment]
Sorry for the delay on this: we've been busy with some typescript changes.

**Comment by mattgodbolt** [comment]
Installing OK but:
```
2023-02-28 03:06:09,966 compilers/dotnet trunk INFO     Most recent dotnet-trunk is 20230227
The user's home directory could not be determined. Set the 'DOTNET_CLI_HOME' environment variable to specify the directory to use.
Installing compilers/dotnet trunk
```
we see a warning. I hope that's OK

**Comment by mattgodbolt** [comment]
The v6.0.11 didn't install:
```
Installing compilers/dotnet v6.0.11
2023-02-28 03:10:38,745 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.11.tar.xz (306741736 bytes)
2023-02-28 03:10:44,065 lib.installation_context INFO     46.5% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.11.tar.xz...
2023-02-28 03:10:49,101 lib.installation_context INFO     95.7% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.11.tar.xz...
2023-02-28 03:10:49,427 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.11.tar.xz
2023-02-28 03:10:49,428 lib.installation_context INFO     Piping to tar Jxf -
2023-02-28 03:13:27,215 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/d477b9cf-0fab-43a8-bca1-052def98bb66/Core_Root) to final destination (/opt/compiler-explorer/dotnet-v6.0.11)
2023-02-28 03:13:44,514 lib.installation_context INFO     Destination /opt/compiler-explorer/dotnet-v6.0.11 exists, temporarily moving out of the way (to /opt/compiler-explorer/staging/d477b9cf-0fab-43a8-bca1-052def98bb66/temp_orig)
The user's home directory could not be determined. Set the 'DOTNET_CLI_HOME' environment variable to specify the directory to use.
2023-02-28 03:14:14,494 lib.ce_install  ERROR    compilers/dotnet v6.0.11 installed OK, but doesn't appear as installed after
```

**Comment by mattgodbolt** [comment]
v6.0.14 failed too:
```
Installing compilers/dotnet v6.0.14
2023-02-28 03:14:14,780 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.14.tar.xz (306775772 bytes)
2023-02-28 03:14:19,805 lib.installation_context INFO     42.4% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.14.tar.xz...
2023-02-28 03:14:25,019 lib.installation_context INFO     93.0% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.14.tar.xz...
2023-02-28 03:14:25,548 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v6.0.14.tar.xz
2023-02-28 03:14:25,549 lib.installation_context INFO     Piping to tar Jxf -
2023-02-28 03:17:02,558 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/f7357b86-3e9e-46cc-b5c1-aaac6a7ed237/Core_Root) to final destination (/opt/compiler-explorer/dotnet-v6.0.14)
2023-02-28 03:17:02,568 lib.installation_context INFO     Directory listing of staging:
total 20
drwxr-xr-x 9 root root 22528 Feb 28 03:17 dotnet-v6.0.14

2023-02-28 03:17:46,895 lib.ce_install  INFO     compilers/dotnet v6.0.14 failed to install: Missing source '/opt/compiler-explorer/staging/f7357b86-3e9e-46cc-b5c1-aaac6a7ed237/Core_Root'
Traceback (most recent call last):
  File "/home/ubuntu/infra/bin/lib/ce_install.py", line 393, in install
    installable.install()
  File "/home/ubuntu/infra/bin/lib/installable/archives.py", line 74, in install
    self.install_context.move_from_staging(staging, self.untar_dir, self.install_path)
  File "/home/ubuntu/infra/bin/lib/installation_context.py", line 202, in move_from_staging
    raise RuntimeError(f"Missing source '{source}'")
RuntimeError: Missing source '/opt/compiler-explorer/staging/f7357b86-3e9e-46cc-b5c1-aaac6a7ed237/Core_Root'

```

**Comment by mattgodbolt** [comment]
Looks like the rest are failing too, unfortunately. Though in fairness I don't know why it's tryig to reinstall 7.0.1 (see below:

```
Installing compilers/dotnet v7.0.1
2023-02-28 03:17:47,168 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v7.0.1.tar.xz (290471908 bytes)
2023-02-28 03:17:52,247 lib.installation_context INFO     53.4% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v7.0.1.tar.xz...
2023-02-28 03:17:56,751 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/dotnet-v7.0.1.tar.xz
2023-02-28 03:17:56,752 lib.installation_context INFO     Piping to tar Jxf -
2023-02-28 03:20:39,613 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/bba0919b-aa7d-48d0-aa94-c8116d4407de/Core_Root) to final destination (/opt/compiler-explorer/dotnet-v7.0.1)
2023-02-28 03:20:39,623 lib.installation_context INFO     Directory listing of staging:
total 20
drwxr-xr-x 9 root root 22528 Feb 28 03:20 dotnet-v7.0.1

2023-02-28 03:21:25,706 lib.ce_install  INFO     compilers/dotnet v7.0.1 failed to install: Missing source '/opt/compiler-explorer/staging/bba0919b-aa7d-48d0-aa94-c8116d4407de/Core_Root'
Traceback (most recent call last):
  File "/home/ubuntu/infra/bin/lib/ce_install.py", line 393, in install
    installable.install()
  File "/home/ubuntu/infra/bin/lib/installable/archives.py", line 74, in install
    self.install_context.move_from_staging(staging, self.untar_dir, self.install_path)
  File "/home/ubuntu/infra/bin/lib/installation_context.py", line 202, in move_from_staging
    raise RuntimeError(f"Missing source '{source}'")
RuntimeError: Missing source '/opt/compiler-explorer/staging/bba0919b-aa7d-48d0-aa94-c8116d4407de/Core_Root'
```
)

**Comment by mattgodbolt** [comment]
Reinstating the
```
    check_env:
      - DOTNET_CLI_HOME=/tmp
```
fixed the "trying to reinstall all the old versions" - the check for "is it installed already?" was failing without this environment variable.

**Comment by mattgodbolt** [comment]
applied some fixes...now installing ok...I hope :)

**Comment by mattgodbolt** [comment]
And installed ok!

**Pull request merged**